File: src/resolve.c
#include "resolve.h"
#include "util.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#ifndef __EMSCRIPTEN__
#include <curl/curl.h>
#endif

// Returns the GitHub base URL (no trailing slash). Defaults to
// "https://github.com" but can be overridden via the ROSIE_GITHUB_BASE_URL
// env var. Used by the regression test suite to point rosie at a local
// mock server.
static const char *github_base_url(void) {
    const char *env = getenv("ROSIE_GITHUB_BASE_URL");
    return (env && *env) ? env : "https://github.com";
}

// --- in-memory fetch target ---

typedef struct {
    char *data;
    size_t size;
    size_t capacity;
} MemBuf;

#ifndef __EMSCRIPTEN__
static size_t mem_write_cb(void *contents, size_t size, size_t nmemb, void *userp) {
    MemBuf *buf = (MemBuf *)userp;
    size_t chunk = size * nmemb;
    if (buf->size + chunk + 1 > buf->capacity) {
        size_t new_cap = buf->capacity == 0 ? 8192 : buf->capacity * 2;
        while (new_cap < buf->size + chunk + 1) new_cap *= 2;
        buf->data = spm_realloc(buf->data, new_cap);
        buf->capacity = new_cap;
    }
    memcpy(buf->data + buf->size, contents, chunk);
    buf->size += chunk;
    buf->data[buf->size] = '\0';
    return chunk;
}

// Fetch the smart-HTTP refs advertisement. Returns 0 on success.
// On success, *out is filled and the caller frees out->data.
static int fetch_refs(const char *owner, const char *repo, MemBuf *out) {
    char url[1024];
    snprintf(url, sizeof(url),
             "%s/%s/%s/info/refs?service=git-upload-pack",
             github_base_url(), owner, repo);

    CURL *curl = curl_easy_init();
    if (!curl) return -1;

    struct curl_slist *headers = NULL;
    headers = curl_slist_append(headers,
        "Accept: application/x-git-upload-pack-advertisement");

    out->data = NULL;
    out->size = 0;
    out->capacity = 0;

    log_debug("Fetching refs: %s", url);
    curl_easy_setopt(curl, CURLOPT_URL, url);
    curl_easy_setopt(curl, CURLOPT_HTTPHEADER, headers);
    curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, mem_write_cb);
    curl_easy_setopt(curl, CURLOPT_WRITEDATA, out);
    curl_easy_setopt(curl, CURLOPT_FOLLOWLOCATION, 1L);
    // Some servers gate smart-HTTP on a git-shaped User-Agent.
    curl_easy_setopt(curl, CURLOPT_USERAGENT, "git/rosie-1.0");
    curl_easy_setopt(curl, CURLOPT_FAILONERROR, 1L);

    if (g_verbose) {
        curl_easy_setopt(curl, CURLOPT_VERBOSE, 1L);
    }

    CURLcode res = curl_easy_perform(curl);
    curl_slist_free_all(headers);
    curl_easy_cleanup(curl);

    if (res != CURLE_OK) {
        log_debug("info/refs fetch failed: %s", curl_easy_strerror(res));
        spm_free(out->data);
        out->data = NULL;
        return -1;
    }
    return 0;
}
#else
// WASM: dispatches through wasm_fetch_to_buffer (implemented in wasm/http-lib.js).
extern int wasm_fetch_to_buffer(const char *url, const char *accept_header,
                                char **out_buf, size_t *out_len);

static int fetch_refs(const char *owner, const char *repo, MemBuf *out) {
    char url[1024];
    snprintf(url, sizeof(url),
             "%s/%s/%s/info/refs?service=git-upload-pack",
             github_base_url(), owner, repo);

    out->data = NULL;
    out->size = 0;
    out->capacity = 0;

    log_debug("Fetching refs: %s", url);
    int status = wasm_fetch_to_buffer(url,
        "application/x-git-upload-pack-advertisement",
        &out->data, &out->size);
    if (status < 0) {
        log_debug("info/refs fetch failed: transport error");
        return -1;
    }
    if (status >= 400) {
        log_debug("info/refs fetch failed: HTTP %d", status);
        return -1;
    }
    out->capacity = out->size;
    return 0;
}
#endif // __EMSCRIPTEN__

// --- pkt-line parsing ---

typedef struct {
    char sha[41];   // null-terminated 40-char hex
    char *name;     // refs/heads/foo, refs/tags/v1.0.0, refs/tags/v1.0.0^{}
} RawRef;

typedef struct {
    RawRef *refs;
    int count;
    int capacity;
} RawRefList;

static void raw_ref_list_free(RawRefList *list) {
    if (!list) return;
    for (int i = 0; i < list->count; i++) {
        spm_free(list->refs[i].name);
    }
    spm_free(list->refs);
    spm_free(list);
}

static int hex_value(char c) {
    if (c >= '0' && c <= '9') return c - '0';
    if (c >= 'a' && c <= 'f') return c - 'a' + 10;
    if (c >= 'A' && c <= 'F') return c - 'A' + 10;
    return -1;
}

static int parse_pkt_len(const char *p) {
    int v = 0;
    for (int i = 0; i < 4; i++) {
        int h = hex_value(p[i]);
        if (h < 0) return -1;
        v = v * 16 + h;
    }
    return v;
}

static RawRefList *parse_refs(const char *body, size_t body_len) {
    RawRefList *list = spm_malloc(sizeof(RawRefList));
    list->refs = NULL;
    list->count = 0;
    list->capacity = 0;

    size_t pos = 0;
    while (pos + 4 <= body_len) {
        int len = parse_pkt_len(body + pos);
        if (len < 0) {
            log_debug("Malformed pkt-line at offset %zu", pos);
            raw_ref_list_free(list);
            return NULL;
        }
        if (len == 0) {
            // Flush packet
            pos += 4;
            continue;
        }
        if (len < 4 || pos + (size_t)len > body_len) {
            log_debug("Bad pkt length %d at offset %zu", len, pos);
            raw_ref_list_free(list);
            return NULL;
        }

        const char *data = body + pos + 4;
        size_t data_len = (size_t)len - 4;
        pos += (size_t)len;

        // Strip trailing newline.
        while (data_len > 0 && (data[data_len - 1] == '\n' || data[data_len - 1] == '\r')) {
            data_len--;
        }

        // Skip the service header line.
        if (data_len > 0 && data[0] == '#') continue;

        // Capabilities follow a NUL byte on the first ref line.
        size_t effective = data_len;
        for (size_t i = 0; i < data_len; i++) {
            if (data[i] == '\0') {
                effective = i;
                break;
            }
        }

        // Format: "<40-hex-sha> <ref-name>"
        if (effective < 42) continue;
        if (data[40] != ' ') continue;
        bool sha_ok = true;
        for (int i = 0; i < 40; i++) {
            if (hex_value(data[i]) < 0) { sha_ok = false; break; }
        }
        if (!sha_ok) continue;

        if (list->count >= list->capacity) {
            int new_cap = list->capacity == 0 ? 32 : list->capacity * 2;
            list->refs = spm_realloc(list->refs, (size_t)new_cap * sizeof(RawRef));
            list->capacity = new_cap;
        }
        RawRef *r = &list->refs[list->count++];
        memcpy(r->sha, data, 40);
        r->sha[40] = '\0';
        size_t name_len = effective - 41;
        r->name = spm_malloc(name_len + 1);
        memcpy(r->name, data + 41, name_len);
        r->name[name_len] = '\0';
    }

    return list;
}

// --- semver parsing ---

typedef struct {
    int major, minor, patch;
    bool has_prerelease;
} SemVer;

// Accept "1.2.3", "v1.2.3", optionally followed by "-..." (prerelease) or "+..." (build).
// Reject anything else (e.g. "1.2", "v1", "release-2026").
static int parse_semver(const char *s, SemVer *out) {
    if (!s || !out) return -1;
    if (s[0] == 'v' || s[0] == 'V') s++;

    out->major = out->minor = out->patch = 0;
    out->has_prerelease = false;

    char *end;
    long n;

    n = strtol(s, &end, 10);
    if (end == s || *end != '.') return -1;
    out->major = (int)n;
    s = end + 1;

    n = strtol(s, &end, 10);
    if (end == s || *end != '.') return -1;
    out->minor = (int)n;
    s = end + 1;

    n = strtol(s, &end, 10);
    if (end == s) return -1;
    out->patch = (int)n;

    if (*end == '\0') return 0;
    if (*end == '-') { out->has_prerelease = true; return 0; }
    if (*end == '+') return 0;
    return -1;
}

static int semver_cmp(const SemVer *a, const SemVer *b) {
    if (a->major != b->major) return a->major - b->major;
    if (a->minor != b->minor) return a->minor - b->minor;
    if (a->patch != b->patch) return a->patch - b->patch;
    // Prereleases sort below their corresponding release (semver §11).
    if (a->has_prerelease != b->has_prerelease) return a->has_prerelease ? -1 : 1;
    return 0;
}

// --- helpers ---

// For an annotated tag, the peeled commit SHA is exposed as "<ref>^{}".
// Return the peeled SHA if present, else the tag's own SHA.
static const char *peeled_sha_for(const RawRefList *refs, int tag_idx) {
    char peeled[512];
    snprintf(peeled, sizeof(peeled), "%s^{}", refs->refs[tag_idx].name);
    for (int i = 0; i < refs->count; i++) {
        if (strcmp(refs->refs[i].name, peeled) == 0) {
            return refs->refs[i].sha;
        }
    }
    return refs->refs[tag_idx].sha;
}

// --- public API ---

ResolvedRef *resolve_latest_tag(const PackageSpec *spec) {
    if (!spec) return NULL;

    MemBuf body;
    if (fetch_refs(spec->owner, spec->repo, &body) != 0) {
        return NULL;
    }
    RawRefList *refs = parse_refs(body.data, body.size);
    spm_free(body.data);
    if (!refs) return NULL;

    int best = -1;
    SemVer best_sv = {0};
    const char *prefix = "refs/tags/";
    size_t prefix_len = strlen(prefix);

    for (int i = 0; i < refs->count; i++) {
        const char *name = refs->refs[i].name;
        if (strncmp(name, prefix, prefix_len) != 0) continue;
        const char *tag = name + prefix_len;
        // Skip peeled refs ("^{}"); we'll look those up explicitly.
        size_t tag_len = strlen(tag);
        if (tag_len >= 3 && strcmp(tag + tag_len - 3, "^{}") == 0) continue;

        SemVer sv;
        if (parse_semver(tag, &sv) != 0) continue;
        if (sv.has_prerelease) continue;

        if (best < 0 || semver_cmp(&sv, &best_sv) > 0) {
            best = i;
            best_sv = sv;
        }
    }

    if (best < 0) {
        raw_ref_list_free(refs);
        return NULL;
    }

    ResolvedRef *result = spm_malloc(sizeof(ResolvedRef));
    result->ref = str_dup(refs->refs[best].name + prefix_len);
    result->sha = str_dup(peeled_sha_for(refs, best));
    result->is_tag = true;

    raw_ref_list_free(refs);
    return result;
}

ResolvedRef *resolve_ref(const PackageSpec *spec, const char *ref_name) {
    if (!spec || !ref_name) return NULL;

    MemBuf body;
    if (fetch_refs(spec->owner, spec->repo, &body) != 0) {
        return NULL;
    }
    RawRefList *refs = parse_refs(body.data, body.size);
    spm_free(body.data);
    if (!refs) return NULL;

    char branch_path[512], tag_path[512];
    snprintf(branch_path, sizeof(branch_path), "refs/heads/%s", ref_name);
    snprintf(tag_path, sizeof(tag_path), "refs/tags/%s", ref_name);

    int found = -1;
    bool is_tag = false;
    for (int i = 0; i < refs->count; i++) {
        if (strcmp(refs->refs[i].name, branch_path) == 0) {
            found = i;
            is_tag = false;
            break;
        }
    }
    if (found < 0) {
        for (int i = 0; i < refs->count; i++) {
            if (strcmp(refs->refs[i].name, tag_path) == 0) {
                found = i;
                is_tag = true;
                break;
            }
        }
    }

    if (found < 0) {
        raw_ref_list_free(refs);
        return NULL;
    }

    ResolvedRef *result = spm_malloc(sizeof(ResolvedRef));
    result->ref = str_dup(ref_name);
    result->sha = str_dup(is_tag ? peeled_sha_for(refs, found) : refs->refs[found].sha);
    result->is_tag = is_tag;

    raw_ref_list_free(refs);
    return result;
}

void resolved_ref_free(ResolvedRef *r) {
    if (!r) return;
    spm_free(r->ref);
    spm_free(r->sha);
    spm_free(r);
}

File: src/resolve.rs
// Git smart-HTTP info/refs resolver. Direct port of resolve.c.
//
// Queries `<base>/<owner>/<repo>/info/refs?service=git-upload-pack`, parses
// the pkt-line response, and exposes:
//   - resolve_auto(spec):          latest semver tag, falling back to the
//                                  default branch (main, then master)
//   - resolve_latest_tag(spec):    highest semver tag (skipping pre-releases)
//   - resolve_ref(spec, ref_name): SHA for a specific branch or tag name
//
// Used by install_package's auto-pin behavior (no @ref → resolve_auto) and by
// `rosie update` to refresh SHAs.

use crate::download::PackageSpec;

#[derive(Debug, Clone)]
pub struct ResolvedRef {
    pub ref_: String,
    pub sha: String,
    pub is_tag: bool,
}

#[derive(Debug, Clone)]
struct RawRef {
    sha: String, // 40-char hex
    name: String,
}

// ---- pkt-line parser ------------------------------------------------------

fn hex_value(c: u8) -> i32 {
    match c {
        b'0'..=b'9' => (c - b'0') as i32,
        b'a'..=b'f' => (c - b'a' + 10) as i32,
        b'A'..=b'F' => (c - b'A' + 10) as i32,
        _ => -1,
    }
}

fn parse_pkt_len(bytes: &[u8]) -> Option<usize> {
    if bytes.len() < 4 {
        return None;
    }
    let mut v: usize = 0;
    for &b in &bytes[..4] {
        let h = hex_value(b);
        if h < 0 {
            return None;
        }
        v = v * 16 + h as usize;
    }
    Some(v)
}

fn parse_refs(body: &[u8]) -> Option<Vec<RawRef>> {
    let mut out: Vec<RawRef> = Vec::new();
    let mut pos = 0;
    while pos + 4 <= body.len() {
        let len = match parse_pkt_len(&body[pos..]) {
            Some(n) => n,
            None => {
                crate::log::debug(&format!("Malformed pkt-line at offset {pos}"));
                return None;
            }
        };
        if len == 0 {
            pos += 4;
            continue;
        }
        if len < 4 || pos + len > body.len() {
            crate::log::debug(&format!("Bad pkt length {len} at offset {pos}"));
            return None;
        }
        let data_start = pos + 4;
        let data_end = pos + len;
        let mut data = &body[data_start..data_end];
        pos = data_end;

        // Strip trailing CR/LF.
        while let Some(&last) = data.last() {
            if last == b'\n' || last == b'\r' {
                data = &data[..data.len() - 1];
            } else {
                break;
            }
        }

        // Service header lines start with '#'.
        if data.first() == Some(&b'#') {
            continue;
        }

        // First ref line: "<sha> <name>\0<capabilities>"
        let effective = data.iter().position(|&b| b == 0).unwrap_or(data.len());
        let data = &data[..effective];

        if data.len() < 42 || data[40] != b' ' {
            continue;
        }
        let sha_ok = data[..40].iter().all(|&b| hex_value(b) >= 0);
        if !sha_ok {
            continue;
        }
        let sha = std::str::from_utf8(&data[..40]).ok()?.to_string();
        let name = std::str::from_utf8(&data[41..]).ok()?.to_string();
        out.push(RawRef { sha, name });
    }
    Some(out)
}

// ---- semver --------------------------------------------------------------

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
struct SemVer {
    major: i32,
    minor: i32,
    patch: i32,
    has_prerelease: bool,
}

/// Accept "1.2.3", "v1.2.3", or 2-part "1.2"/"v1.2" (treated as "1.2.0"),
/// optionally followed by "-..." (prerelease) or "+..." (build). Reject
/// anything else (e.g. "v1", "release-2026").
fn parse_semver(s: &str) -> Option<SemVer> {
    let mut s = s.as_bytes();
    if matches!(s.first(), Some(b'v') | Some(b'V')) {
        s = &s[1..];
    }
    fn take_num(s: &[u8]) -> Option<(i32, &[u8])> {
        let mut i = 0;
        while i < s.len() && s[i].is_ascii_digit() {
            i += 1;
        }
        if i == 0 {
            return None;
        }
        let n = std::str::from_utf8(&s[..i]).ok()?.parse::<i32>().ok()?;
        Some((n, &s[i..]))
    }

    let (major, rest) = take_num(s)?;
    if rest.first() != Some(&b'.') {
        return None;
    }
    let (minor, rest) = take_num(&rest[1..])?;
    // Optional patch — accept 2-part "X.Y" as "X.Y.0" so tags like
    // "v0.5" sort correctly against "v0.2.1".
    let (patch, rest) = if rest.first() == Some(&b'.') {
        take_num(&rest[1..])?
    } else {
        (0, rest)
    };
    let has_prerelease = match rest.first() {
        None => false,
        Some(&b'-') => true,
        Some(&b'+') => false,
        _ => return None,
    };
    Some(SemVer {
        major,
        minor,
        patch,
        has_prerelease,
    })
}

fn semver_cmp(a: &SemVer, b: &SemVer) -> std::cmp::Ordering {
    a.major
        .cmp(&b.major)
        .then(a.minor.cmp(&b.minor))
        .then(a.patch.cmp(&b.patch))
        // Prereleases sort below their corresponding release (semver §11).
        .then(b.has_prerelease.cmp(&a.has_prerelease))
}

fn peeled_sha_for(refs: &[RawRef], tag_idx: usize) -> &str {
    let peeled = format!("{}^{{}}", refs[tag_idx].name);
    for r in refs {
        if r.name == peeled {
            return &r.sha;
        }
    }
    &refs[tag_idx].sha
}

// ---- public API ----------------------------------------------------------

fn fetch_refs(owner: &str, repo: &str) -> Option<Vec<u8>> {
    let base = crate::http::github_base_url();
    let url = format!("{base}/{owner}/{repo}/info/refs?service=git-upload-pack");
    let (status, body) = crate::http::fetch_to_buffer(
        &url,
        Some("application/x-git-upload-pack-advertisement"),
    );
    if status < 0 {
        crate::log::debug("info/refs fetch failed: transport error");
        return None;
    }
    if status >= 400 {
        crate::log::debug(&format!("info/refs fetch failed: HTTP {status}"));
        return None;
    }
    Some(body)
}

/// Walk `refs` and return the highest semver tag matching one of `tag_prefixes`
/// (each prefix should already include "refs/tags/..."). Pre-releases skipped.
fn best_tag<'a, I>(refs: &[RawRef], tag_prefixes: I) -> Option<(usize, SemVer)>
where
    I: IntoIterator<Item = &'a str>,
{
    let prefixes: Vec<&str> = tag_prefixes.into_iter().collect();
    let mut best: Option<(usize, SemVer)> = None;
    for (i, r) in refs.iter().enumerate() {
        let rest = match prefixes.iter().find_map(|p| r.name.strip_prefix(*p)) {
            Some(t) => t,
            None => continue,
        };
        if rest.ends_with("^{}") {
            continue;
        }
        let sv = match parse_semver(rest) {
            Some(v) => v,
            None => continue,
        };
        if sv.has_prerelease {
            continue;
        }
        match &best {
            None => best = Some((i, sv)),
            Some((_, b)) if semver_cmp(&sv, b).is_gt() => best = Some((i, sv)),
            _ => {}
        }
    }
    best
}

fn pick_latest_tag(refs: &[RawRef], repo: &str) -> Option<ResolvedRef> {
    // Try name-prefixed monorepo conventions first (e.g. "react-doctor@0.0.38",
    // "foo-v1.2.3"). Fall back to bare-semver tags (e.g. "v0.5").
    let scoped_at = format!("refs/tags/{repo}@");
    let scoped_dashv = format!("refs/tags/{repo}-v");
    let scoped = [scoped_at.as_str(), scoped_dashv.as_str()];
    let bare = ["refs/tags/"];

    let (idx, _) = best_tag(refs, scoped.iter().copied())
        .or_else(|| best_tag(refs, bare.iter().copied()))?;

    let name = refs[idx].name.clone();
    let tag = name.strip_prefix("refs/tags/").unwrap_or(&name).to_string();
    Some(ResolvedRef {
        ref_: tag,
        sha: peeled_sha_for(refs, idx).to_string(),
        is_tag: true,
    })
}

fn pick_branch(refs: &[RawRef], name: &str) -> Option<ResolvedRef> {
    let path = format!("refs/heads/{name}");
    refs.iter().find(|r| r.name == path).map(|r| ResolvedRef {
        ref_: name.to_string(),
        sha: r.sha.clone(),
        is_tag: false,
    })
}

pub fn resolve_latest_tag(spec: &PackageSpec) -> Option<ResolvedRef> {
    let owner = spec.owner.as_deref()?;
    let repo = spec.repo.as_deref()?;
    let body = fetch_refs(owner, repo)?;
    let refs = parse_refs(&body)?;
    pick_latest_tag(&refs, repo)
}

/// Auto-pin resolution: prefer the highest semver tag, otherwise fall back to
/// the default branch ("main", then "master"). Fetches `info/refs` once and
/// inspects all candidates from that single response.
pub fn resolve_auto(spec: &PackageSpec) -> Option<ResolvedRef> {
    let owner = spec.owner.as_deref()?;
    let repo = spec.repo.as_deref()?;
    let body = fetch_refs(owner, repo)?;
    let refs = parse_refs(&body)?;

    if let Some(r) = pick_latest_tag(&refs, repo) {
        return Some(r);
    }
    pick_branch(&refs, "main").or_else(|| pick_branch(&refs, "master"))
}

pub fn resolve_ref(spec: &PackageSpec, ref_name: &str) -> Option<ResolvedRef> {
    let owner = spec.owner.as_deref()?;
    let repo = spec.repo.as_deref()?;
    let body = fetch_refs(owner, repo)?;
    let refs = parse_refs(&body)?;

    let branch_path = format!("refs/heads/{ref_name}");
    let tag_path = format!("refs/tags/{ref_name}");

    if let Some((i, _)) = refs.iter().enumerate().find(|(_, r)| r.name == branch_path) {
        return Some(ResolvedRef {
            ref_: ref_name.to_string(),
            sha: refs[i].sha.clone(),
            is_tag: false,
        });
    }
    if let Some((i, _)) = refs.iter().enumerate().find(|(_, r)| r.name == tag_path) {
        return Some(ResolvedRef {
            ref_: ref_name.to_string(),
            sha: peeled_sha_for(&refs, i).to_string(),
            is_tag: true,
        });
    }
    None
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn semver_accepts_v_prefix() {
        let a = parse_semver("v1.2.3").unwrap();
        assert_eq!(
            a,
            SemVer {
                major: 1,
                minor: 2,
                patch: 3,
                has_prerelease: false
            }
        );
    }

    #[test]
    fn semver_rejects_partial() {
        assert!(parse_semver("v1").is_none());
        assert!(parse_semver("release-2026").is_none());
    }

    #[test]
    fn semver_accepts_two_part() {
        let two = parse_semver("v0.5").unwrap();
        assert_eq!(two, SemVer { major: 0, minor: 5, patch: 0, has_prerelease: false });
        // "0.5" sorts above "0.2.1"
        let old = parse_semver("v0.2.1").unwrap();
        assert!(semver_cmp(&two, &old).is_gt());
    }

    #[test]
    fn semver_orders_prerelease_below() {
        let release = parse_semver("1.0.0").unwrap();
        let pre = parse_semver("1.0.0-rc.1").unwrap();
        assert!(semver_cmp(&pre, &release).is_lt());
    }

    #[test]
    fn pkt_line_smoke() {
        // Build a minimal pkt-line stream: service header, flush, ref line, flush.
        fn pkt(payload: &[u8]) -> Vec<u8> {
            let mut out = format!("{:04x}", payload.len() + 4).into_bytes();
            out.extend_from_slice(payload);
            out
        }
        let mut body = pkt(b"# service=git-upload-pack\n");
        body.extend_from_slice(b"0000");
        body.extend_from_slice(&pkt(
            b"1111111111111111111111111111111111111111 refs/heads/main\0caps\n",
        ));
        body.extend_from_slice(&pkt(b"aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa refs/tags/v1.0.0\n"));
        body.extend_from_slice(b"0000");

        let refs = parse_refs(&body).unwrap();
        // Only the two real refs; service header skipped.
        assert_eq!(refs.len(), 2);
        assert_eq!(refs[0].name, "refs/heads/main");
        assert_eq!(refs[0].sha, "1".repeat(40));
        assert_eq!(refs[1].name, "refs/tags/v1.0.0");
    }

    fn raw(sha_byte: char, name: &str) -> RawRef {
        RawRef {
            sha: std::iter::repeat(sha_byte).take(40).collect(),
            name: name.to_string(),
        }
    }

    #[test]
    fn pick_tag_two_part_beats_three_part() {
        // jdevalk/skills shape: v0.5 must win over v0.2.1.
        let refs = vec![
            raw('1', "refs/tags/v0.1"),
            raw('2', "refs/tags/v0.2"),
            raw('3', "refs/tags/v0.2.1"),
            raw('4', "refs/tags/v0.3"),
            raw('5', "refs/tags/v0.4"),
            raw('6', "refs/tags/v0.5"),
        ];
        let r = pick_latest_tag(&refs, "skills").unwrap();
        assert_eq!(r.ref_, "v0.5");
        assert!(r.is_tag);
    }

    #[test]
    fn pick_tag_prefers_repo_scoped() {
        // millionco/react-doctor shape: prefer "react-doctor@0.0.38"
        // over the lone bare "0.0.1".
        let refs = vec![
            raw('1', "refs/tags/0.0.1"),
            raw('2', "refs/tags/react-doctor@0.0.31"),
            raw('3', "refs/tags/react-doctor@0.0.38"),
        ];
        let r = pick_latest_tag(&refs, "react-doctor").unwrap();
        assert_eq!(r.ref_, "react-doctor@0.0.38");
    }

    #[test]
    fn pick_tag_repo_scoped_dash_v() {
        // <repo>-v<ver> convention (some Go monorepos).
        let refs = vec![
            raw('1', "refs/tags/v9.9.9"), // bare; should lose to scoped.
            raw('2', "refs/tags/foo-v1.0.0"),
            raw('3', "refs/tags/foo-v1.2.0"),
        ];
        let r = pick_latest_tag(&refs, "foo").unwrap();
        assert_eq!(r.ref_, "foo-v1.2.0");
    }

    #[test]
    fn pick_tag_falls_back_to_bare_when_no_scoped() {
        let refs = vec![
            raw('1', "refs/tags/v0.2.1"),
            raw('2', "refs/tags/v0.5"),
        ];
        let r = pick_latest_tag(&refs, "anything").unwrap();
        assert_eq!(r.ref_, "v0.5");
    }

    #[test]
    fn pick_tag_skips_prereleases() {
        let refs = vec![
            raw('1', "refs/tags/v1.0.0"),
            raw('2', "refs/tags/v1.1.0-rc.1"),
        ];
        let r = pick_latest_tag(&refs, "x").unwrap();
        assert_eq!(r.ref_, "v1.0.0");
    }

    #[test]
    fn pick_tag_none_when_no_tags() {
        let refs = vec![raw('1', "refs/heads/main")];
        assert!(pick_latest_tag(&refs, "x").is_none());
    }

    #[test]
    fn pick_branch_prefers_main_then_master() {
        let with_main = vec![
            raw('a', "refs/heads/main"),
            raw('b', "refs/heads/master"),
        ];
        let r = pick_branch(&with_main, "main").unwrap();
        assert_eq!(r.ref_, "main");
        assert!(!r.is_tag);
        assert_eq!(r.sha.chars().next(), Some('a'));

        // master-only repo
        let master_only = vec![raw('c', "refs/heads/master")];
        assert!(pick_branch(&master_only, "main").is_none());
        let r = pick_branch(&master_only, "master").unwrap();
        assert_eq!(r.ref_, "master");
    }
}
